core/src/main/java/org/lflang/analyses/dag/Dag.java

```diff
@@ -0,0 +1,319 @@
+      HashMap<DagNode, DagEdge> inner = this.dagEdges.get(source);
+      if (inner != null) {
+        for (DagNode sink : inner.keySet()) {
+          int sourceIdx = dagNodes.indexOf(source);
+          int sinkIdx = dagNodes.indexOf(sink);
+          dot.pr(sourceIdx + " -> " + sinkIdx);
+        }
+      }
+    }
+
+    dot.unindent();
+    dot.pr("}");
+
+    return this.dot;
+  }
+
+  public void generateDotFile(Path filepath) {
+    try {
+      CodeBuilder dot = generateDot();
+      String filename = filepath.toString();
+      dot.writeToFile(filename);
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  /**
+   * Parses the dot file, reads the edges and updates the DAG. We assume that the edges are
+   * specified as: <SrcNodeId> -> <SinkNodeId>.
+   *
+   * <p>Furthermore, we assume that the new DAG (contained in the dotFile) will not have unnecessary
+   * edges, since they are removed by the shceduler.
+   *
+   * @param dotFilename
+   * @return
+   */
+  public boolean updateDag(String dotFileName) throws IOException {
+    FileReader fileReader;
+    BufferedReader bufferedReader;
+    // Read the file
+    try {
+      fileReader = new FileReader(dotFileName);
+      // Buffer the input stream from the file
+      bufferedReader = new BufferedReader(fileReader);
+    } catch (IOException e) {
+      System.out.println("Problem accessing file " + dotFileName + "! " + e);
+      return false;
+    }
+
+    String line;
+
+    // Pattern with which an edge starts:
+    Pattern pattern = Pattern.compile("^((\s*)(\\d+)(\s*)->(\s*)(\\d+))");
+    Matcher matcher;
+
+    // Before iterating to search for the edges, we clear the DAG edges array list
+    this.dagEdges.clear();
+
+    // Search
+    int i = 0;
+    while (bufferedReader.ready()) {
+
+      line = bufferedReader.readLine();
+      matcher = pattern.matcher(line);
+      if (matcher.find()) {
+        // This line describes an edge
+        // Start by removing all white spaces. Only the nodes ids and the
+        // arrow remain in the string.
+        line = line.replaceAll("\\s", "");
+
+        // Remove the label and the ';' that may appear after the edge specification
+        StringTokenizer st = new StringTokenizer(line, ";");
+        line = st.nextToken();
+        st = new StringTokenizer(line, "[");
+        line = st.nextToken();
+
+        // Use a StringTokenizer to find the source and sink nodes' ids
+        st = new StringTokenizer(line, "->");
+        int srcNodeId, sinkNodeId;
+
+        // Get the source and sink nodes ids and add the edge
+        try {
+          srcNodeId = Integer.parseInt(st.nextToken());
+          sinkNodeId = Integer.parseInt(st.nextToken());
+          this.addEdge(srcNodeId, sinkNodeId);
+        } catch (NumberFormatException e) {
+          System.out.println("Parse error in line " + line + " : Expected a number!");
+          Exceptions.sneakyThrow(e);
+        }
```

Comment:
Actually it occurs to me that this whole int extraction logic could be replaced by using the capture groups of the matcher. Consider rewriting your pattern
```java
"^\\s*(\\d+)\\s*->\\s*(\\d+)"
```
to remove useless capture groups. Notice also that we use `\\s` and not `\s` which is a string escape and not what we want. Then after `matcher.find()` all you need is to write
```java
int srcNodeId = Integer.parseInt(matcher.group(1));
int sinkNodeId = Integer.parseInt(matcher.group(2));
this.addEdge(srcNodeId, sinkNodeId);
```
This also makes it clear that what you're giving to your `Integer.parseInt` is a sequence of digits so you don't need to catch a NumberFormatException, that will never occur.